examples/nlp/semantic_similarity_with_bert.py

```diff
@@ -0,0 +1,363 @@
+Date created: 2020/08/15
+Last modified: 2020/08/20
+Description: Natural Language Inference by Fine-tuning BERT model on SNLI Corpus.
+"""
+"""
+## Introduction
+
+Semantic Similarity or Natural Languauge Inference is the task of determining how two
+sentences are similar to each other in terms of their meaning.
+This example demonstrates the use of SNLI (Standford Natural Language Inference) Corpus
+to perform semantic similarity with Transformers.
+We will fine-tune BERT model by providing two sentences as inputs and it outputs the
+probability of the similarity between sentences.
+
+### References
+* [BERT](https://arxiv.org/pdf/1810.04805.pdf)
+* [SNLI](https://nlp.stanford.edu/projects/snli/)
+"""
+
+"""
+## Setup
+"""
+
+import numpy as np
+import pandas as pd
+import tensorflow as tf
+import transformers
+from transformers import BertTokenizer, TFBertModel
+
+"""
+## Configurations
+"""
+
+max_length = 128  # Maximum length of input sentence to the model.
+batch_size = 32
+epochs = 4
+learning_rate = 3e-5
+np.random.seed(42)
+
+"""
+## Load the Data
+"""
+
+"""shell
+curl -LO https://raw.githubusercontent.com/MohamadMerchant/SNLI/master/data.tar.gz
+tar -xvzf data.tar.gz
+"""
+
+train_df = pd.read_csv("SNLI_Corpus/snli_1.0_train.csv")
+valid_df = pd.read_csv("SNLI_Corpus/snli_1.0_dev.csv")
+test_df = pd.read_csv("SNLI_Corpus/snli_1.0_test.csv")
+
+# Shape of the data
+print(f"Total train samples : {train_df.shape[0]}")
+print(f"Total validation samples: {valid_df.shape[0]}")
+print(f"Total test samples: {valid_df.shape[0]}")
+
+"""
+Dataset Overview:
+
+sentence1: The premise caption that was supplied to the author of the pair.
+
+sentence2: The hypothesis caption that was written by the author of the pair.
+
+similarity: This is the label chosen by the majority of annotators.
+            Where no majority exists, this is '-' which we will not use in our task.
+
+Types of similarity/labels in our dataset:
+Contradiction: The sentences have no similarity or different from each other.
+Entailment:    The sentences have similar meaning.
+Neutral:       The sentences are neutral.
+"""
+
+"""
+Lets look at one sample of data:
+"""
+print(f"Sentence1: {train_df.loc[1, 'sentence1']}")
+print(f"Sentence2: {train_df.loc[1, 'sentence2']}")
+print(f"Similarity: {train_df.loc[1, 'similarity']}")
+
+"""
+## Preprocessing
+"""
+
+# We have some nan in our train data, we will simply drop them.
+print(train_df.isnull().sum())
+train_df.dropna(axis=0, inplace=True)
+
+# We have some "-" in our train and validation targets, so we will not use those.
+train_df = (
+    train_df[train_df.similarity != "-"]
+    .sample(frac=1.0, random_state=42)
+    .reset_index(drop=True)
+)
+valid_df = (
+    valid_df[valid_df.similarity != "-"]
+    .sample(frac=1.0, random_state=42)
+    .reset_index(drop=True)
+)
+
+"""
+Distribution of our training and validation targets.
+"""
+print(train_df.similarity.value_counts())
```

Comment:
Use string formatting to print a sentence that makes it clear this is training (resp. validation) data